baseplate/lib/circuit_breaker/breaker.py

```diff
@@ -0,0 +1,127 @@
+from collections import deque
+from datetime import datetime
+from datetime import timedelta
+from enum import Enum
+from math import ceil
+from random import random
+from typing import Deque
+
+
+class BreakerState(Enum):
+    WORKING = "working"
+    TRIPPED = "tripped"
+    # trip immediately after failure
+    TESTING = "testing"
+
+
+class Breaker:
+    """Circuit breaker.
+
+    The circuit breaker has 3 states:
+    * WORKING (closed)
+    * TRIPPED (open)
+    * TESTING (half open)
+
+    During normal operation the circuit breaker is in the WORKING state.
+
+    When the number of failures exceeds the threshold the breaker moves to the TRIPPED state. It
+    stays in this state for the timeout period.
+
+    After the timeout period passes the breaker moves to the TESTING state. If the next attempt
+    is successful the breaker moves to the WORKING state. If the next attempt is a failure the
+    breaker moves back to the TRIPPED state.
+
+    :param name: full name/path of the circuit breaker
+    :param samples: number of previous results used to calculate the trip failure ratio
+    :param trip_failure_percent: the minimum ratio of sampled failed results to trip the breaker
+    :param trip_for: how long to remain tripped before resetting the breaker
+    :param fuzz_ratio: how much to randomly add/subtract to the trip_for time
+    """
+
```

Comment:
🔕 My linter complains about variables initialized outside init, feel free to disregard

```suggestion

failures: int
```

Reply:
yeah that's a little funky that it's only initialized in `reset()`(which is called from the constructor), but it does prevent some duplication. I'm leaning towards leaving it as is because this is copied exactly from the gql implementation.

Reply:
it'd be good to add the type annotation just so everything knows it can expect the attribute to be there